boefjes/boefjes/plugins/kat_webpage_capture/main.py

```diff
@@ -68,51 +78,63 @@ def get_file_from_container(container: docker.models.containers.Container, path:
 
 def build_playwright_command(webpage: str, browser: str, tmp_path: str) -> str:
     """Returns playwright command including webpage, browser and locations for image, har and storage."""
-    return " ".join(
-        [
-            "npx playwright screenshot",
-            f"-b {browser}",
-            "--full-page",
-            f"--save-har={tmp_path}.har.zip",
-            f"--save-storage={tmp_path}.json",
-            webpage,
-            f"{tmp_path}.png",
-        ]
-    )
+    return [
+        "/usr/bin/npx",
+        "playwright",
+        "screenshot",
+        "-b",
+        browser,
+        "--full-page",
+        f"--save-har={tmp_path}.har.zip",
+        f"--save-storage={tmp_path}.json",
+        webpage,
+        f"{tmp_path}.png",
+    ]
 
 
 def run_playwright(webpage: str, browser: str, tmp_path: str = "/tmp/tmp") -> Tuple[bytes]:
     """Run Playwright in Docker."""
     client = docker.from_env()
-    res = client.containers.run(
+    client.images.pull(PLAYWRIGHT_IMAGE)
+    # https://playwright.dev/docs/docker#crawling-and-scraping
+    command = build_playwright_command(webpage=webpage, browser=browser, tmp_path=tmp_path)
+    container = client.containers.run(
         image=PLAYWRIGHT_IMAGE,
-        command=[
-            "/bin/sh",
-            "-c",
-            f"npx -y {build_playwright_command(webpage=webpage, browser=browser, tmp_path=tmp_path)}",
-        ],
+        command=command,
         detach=True,
+        ipc_mode="host",
+        user="pwuser",
+        security_opt=[
+            (
+                'seccomp={"comment": "Allow create user namespaces", "names": ["clone", "setns", "unshare"], '
+                '"action": "SCMP_ACT_ALLOW", "args": [], "includes": {}, "excludes": {}}'
+            )
+        ],
     )
     try:
-        res.wait()
-        image = Image.open(io.BytesIO(get_file_from_container(container=res, path=f"{tmp_path}.png")))
-        har = get_file_from_container(container=res, path=f"{tmp_path}.har.zip")
-        storage = get_file_from_container(container=res, path=f"{tmp_path}.json")
+        container.wait()
+        image = get_file_from_container(container=container, path=f"{tmp_path}.png")
+        har = get_file_from_container(container=container, path=f"{tmp_path}.har.zip")
+        storage = get_file_from_container(container=container, path=f"{tmp_path}.json")
+        return image, har, storage
+    except docker.errors.NotFound:
+        raise WebpageCaptureException(
+            "Error while running Playwright container, command was: " + " ".join(command),
+            container.logs(stdout=True, stderr=True, timestamps=True).decode(),
+        )
```

Comment:
Raising this error will result in no files saved when one of the files is actually present. Suppose the screenshot was succesfully taken but for some reason the `har.zip` was not created. Then the screenshot is also not returned anymore. (That is why it was designed to only raise a warning for each missing file)

Reply:
Showing the container logs is a very nice addition by the way. It would be great if that can also be added to other boefjes using containers for their commands (e.g. Nmap).

Reply:
Without this PR the container crashes when the screenshot is not present, it doesn't just log a warning.

Because we don't have a "partially successful" status code for Boefjes, I'd say the only reasonable way to handle this is to fail when any file is missing.

Reply:
> Showing the container logs is a very nice addition by the way. It would be great if that can also be added to other boefjes using containers for their commands (e.g. Nmap).

#1191 already exists for that.

Reply:
> Because we don't have a "partially successful" status code for Boefjes, I'd say the only reasonable way to handle this is to fail when any file is missing.

I created a separate issue https://github.com/minvws/nl-kat-coordination/issues/1440